docs/smtchecker.rst

```diff
@@ -19,22 +19,110 @@ SMTChecker module automatically tries to prove that the code satisfies the
+
+.. note::
+  The lack of warnings for a verification target represents an undisputed
+  mathematical proof of correctness. Keep in mind that these problems are
+  *very hard* and sometimes *impossible* to solve automatically in the
+  general case.  Therefore, several properties might not be solved or might
+  lead to false positives for large contracts. Every proven property should
+  be seen as an important achievement. For advanced users, see `ref: ...`
+  to learn a few SMTChecker options that might help proving more complex
+  properties.
+  
 It is currently an experimental feature, therefore in order to use it you need
 to enable it via :ref:`a pragma directive<smt_checker>`.
 
+***************
+Simple Tutorial
+***************
+
+.. code-block:: Solidity
+
+    pragma experimental SMTChecker;
+
+    contract Overflow {
+        uint immutable x;
+        uint immutable y;
+
+        function add(uint _x, uint _y) internal pure returns (uint) {
+            return _x + _y;
+        }
+
+        constructor(uint _x, uint _y) {
+            (x, y) = (_x, _y);
+        }
+
+        function stateAdd() public view returns (uint) {
+            return add(x, y);
+        }
+    }
+
+The contract above shows an overflow check example.
+The SMTChecker will, by default, check every reachable arithmetic operation
+in the contract for potential underflow and overflow.
+Here, it reports the following:
+
+.. code-block:: bash
+
+    Warning: CHC: Overflow (resulting value larger than 2**256 - 1) happens here.
+    Counterexample:
+    x = 1, y = 115792089237316195423570985008687907853269984665640564039457584007913129639935
+     = 0
+    
+    Transaction trace:
+    Overflow.constructor(1, 115792089237316195423570985008687907853269984665640564039457584007913129639935)
+    State: x = 1, y = 115792089237316195423570985008687907853269984665640564039457584007913129639935
+    Overflow.stateAdd()
+        Overflow.add(1, 115792089237316195423570985008687907853269984665640564039457584007913129639935) -- internal call
+     --> o.sol:9:20:
+      |
+    9 |             return _x + _y;
+      |                    ^^^^^^^
+
+If we add ``require`` statements that filter out overflow cases,
+the SMTChecker proves that no overflow is reachable:
+
+.. code-block:: Solidity
+
+    pragma experimental SMTChecker;
+
+    contract Overflow {
+        uint immutable x;
+        uint immutable y;
+
+        function add(uint _x, uint _y) internal pure returns (uint) {
+            return _x + _y;
+        }
+
+        constructor(uint _x, uint _y) {
+            (x, y) = (_x, _y);
+        }
+
+        function stateAdd() public view returns (uint) {
+            require(x < type(uint128).max);
+            require(y < type(uint128).max);
```

Comment:
`require(x < type(uint248).max)` and `require(y < type(uint248).max)` is a more fun example. Fine with keeping this.

Reply:
Yea I thought about that too, but felt like 128 is more "relatable" or more common. I'm also fine with either.